File: server/execution_environments/python/runner.py
import sys
import json
import numpy
import pandas

def execute():
    try:
        with open('code.py', 'r') as f:
            code = f.read()

        with open('args.json', 'r') as f:
            args = json.load(f)

        # Prepare the execution scope
        globals_dict = {
            '__builtins__': {
                'print': print,
                'len': len,
                'range': range,
                'str': str,
                'int': int,
                'float': float,
                'list': list,
                'dict': dict,
                'tuple': tuple,
                'set': set,
                'abs': abs,
                'round': round,
                'max': max,
                'min': min,
                'sum': sum,
                'sorted': sorted,
                '__import__': __import__
            },
            'numpy': numpy,
            'pandas': pandas,
        }
        locals_dict = {'args': args}

        # Wrap the user code in a function
        wrapped_code = f"""
def NOUVELLE_FONCTION(*args):
{code}

result = NOUVELLE_FONCTION(*args)
"""
        exec(wrapped_code, globals_dict, locals_dict)

        result = locals_dict.get('result')

        # Serialize the result to JSON
        # This is a very basic serializer, might need to be improved
        # to handle more complex types (e.g., numpy arrays)
        def default_serializer(o):
            if isinstance(o, (numpy.ndarray,)):
                return o.tolist()
            if isinstance(o, (numpy.generic,)):
                return o.item()
            return f"unserializable type: {type(o)}"

        output = json.dumps({"result": result}, default=default_serializer)

        print(output)

    except Exception as e:
        error_output = json.dumps({"error": str(e)})
        print(error_output, file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    execute()

File: server/execution_environments/lua/runner.lua
local json = require("dkjson")

local function execute()
    -- Open and read code file
    local f_code = io.open("code.lua", "r")
    if not f_code then
        io.stderr:write(json.encode({error = "Could not open code.lua"}))
        os.exit(1)
    end
    local code = f_code:read("*a")
    f_code:close()

    -- Open and read args file
    local f_args = io.open("args.json", "r")
    if not f_args then
        io.stderr:write(json.encode({error = "Could not open args.json"}))
        os.exit(1)
    end
    local args_json_str = f_args:read("*a")
    f_args:close()

    -- Decode arguments
    local args, pos, err = json.decode(args_json_str)
    if err then
        io.stderr:write(json.encode({error = "JSON decoding error for args: " .. err}))
        os.exit(1)
    end

    -- Wrap code in a function
    local func_str = "return function(...)\n" .. code .. "\nend"
    local func, load_err = load(func_str)
    if not func then
        io.stderr:write(json.encode({error = "Syntax error: " .. load_err}))
        os.exit(1)
    end

    -- Execute the function with protected call
    local status, result = pcall(func(), table.unpack(args))

    if status then
        -- Encode and print result
        local result_json, encode_err = json.encode({result = result})
        if encode_err then
             io.stderr:write(json.encode({error = "JSON encoding error for result: " .. encode_err}))
             os.exit(1)
        end
        print(result_json)
    else
        -- Print error from execution
        io.stderr:write(json.encode({error = "Execution error: " .. tostring(result)}))
        os.exit(1)
    end
end

execute()
